src/relations/database.py

```diff
@@ -43,6 +43,135 @@ def __init__(self, charm):
         self.framework.observe(
             self.charm.on[DB_RELATION_NAME].relation_broken, self._on_database_broken
         )
+        self.framework.observe(self.charm.on[PEER].relation_joined, self._on_relation_joined)
+        self.framework.observe(self.charm.on[PEER].relation_departed, self._on_relation_departed)
+
+        self.framework.observe(self.charm.on.leader_elected, self._on_leader_elected)
+
+    def _on_leader_elected(self, _):
+        """Handle on leader elected event for the database relation."""
+        if not self.charm.unit.is_leader():
+            return
+        # get all relations involving the database relation
+        relations = list(self.model.relations[DB_RELATION_NAME])
+        # check if there are relations in place
+        if len(relations) == 0:
+            return
+
+        if not self.charm.cluster_initialized:
+            logger.debug("Waiting cluster to be initialized")
+            return
+
+        relation_data = self.database.fetch_relation_data()
+        # for all relations update the read-only-endpoints
+        for relation in relations:
+            # check if the on_database_requested has been executed
+            if relation.id not in relation_data:
+                logger.debug("On database requested not happened yet! Nothing to do in this case")
+                continue
+            self._update_endpoints(relation.id, self.charm.unit.name)
+
+    def _on_relation_departed(self, event: RelationDepartedEvent):
+        """Handle the peer relation departed event for the database relation."""
+        if not self.charm.unit.is_leader():
+            return
+        # get all relations involving the database relation
+        relations = list(self.model.relations[DB_RELATION_NAME])
+        if len(relations) == 0:
+            return
+
+        if not self.charm.cluster_initialized:
+            logger.debug("Waiting cluster to be initialized")
+            return
+
+        # check if the leader is departing
+        if self.charm.unit.name == event.departing_unit.name:
+            event.defer()
```

Comment:
do we need to defer here? if the leader is the departing unit, may be we can just return, and let the `leader-elected` handler above handle the update of the endpoints

Reply:
Good catch! I changed it